### PR TITLE
Fix #201: focussearch now handles discrete vector parameters

### LINKED ISSUE #201: FocusSearch doesn't handle vector parameters
In particular, I reliably get an error [here](https://github.com/mlr-org/mlrMBO/blob/7ca9ad456e18024f8d1d1cacfe8f1a3ad3a67360/R/infillOptFocus.R#L58) when I have a discrete vector parameter, since `val` is a list and `setdiff(val.names, val)` can make `val.names` empty even when it was `length(par$values) > 1L` before.

Should I, in general, assume that mlrMBO wasn't designed to handle vector parameters and give up right now, or is this worth a fix?

### PR DESCRIPTION
**Pull request by mb706**
Also general focus search bugfixes.

### CODE REVIEW

File: R/infillOptFocus.R [hunk 1]
```diff
@@ -21,13 +29,22 @@ infillOptFocus = function(infill.crit, models, control, par.set, opt.path, desig
...
+
+      # handle discrete vectors
+      for (dvparam in filterParams(par.set, type = c("discretevector", "logicalvector"))$pars) {
+        for (dimnum in seq_len(dvparam$len)) {
+          dfindex = paste0(dvparam$id, dimnum)
+          mapping = discreteVectorMapping[[dvparam$id]][[dimnum]]
+          levels(newdesign[[dfindex]]) = mapping[levels(newdesign[[dfindex]])]
+        }
+      }
+      y = infill.crit(newdesign, models, control, par.set, design, iter, ...)
```
Comment (jakob-r) [inline R/infillOptFocus.R]: This is potentially slow and doesn't make use of our helpers. have a look at
```{r}
par.set.disc.logic.v = filterParams(par.set, type = c("discretevector", "logicalvector"))
dfindex = getParamIds(par.set.disc.logic.v, with.nr = TRUE, repeated = TRUE)
```

**Comment by mb706** [comment]
Incorporated your suggestions.

I also found out that this cannot work when there is a parameter that has a discrete vector parameter in its requirement. I think for this to work one needs to either rewrite the requirements or change something in ParamHelpers.

**Comment by jakob-r** [comment]
As @berndbischl is the *inventor* of foucssearch he should have a final look on it. If I see somtehing in the code I will commit.

File: R/infillOptFocus.R [hunk 2]
```diff
@@ -8,11 +8,35 @@
 # See infillOptCMAES.R for interface explanation.
 infillOptFocus = function(infill.crit, models, control, par.set, opt.path, design, iter, ...) {
   global.y = Inf
+  
+  allRequirements = extractSubList(par.set$pars, "requires", simplify = FALSE)
+  allUsedVars = unique(do.call(base::c, sapply(allRequirements, all.vars)))
```
Comment (jakob-r) [inline R/infillOptFocus.R]: What is `all.vars` in this context? I feel stupid, but I cannot find it.
I also find it strange to use `do.call(base::c, ...)`. Won't `unlist` do the trick?

File: R/infillOptFocus.R [hunk 3]
```diff
@@ -8,11 +8,35 @@
...
     ps.local = par.set
+    
+    # handle discrete vectors:
+    # The problem is that for discrete vectors, we can't adjust the values dimension-wise.
+    # Therefore, for discrete vectors, we always drop the last level and instead have a
+    # mapping that maps, for each discrete vector param and for each dimension, from 
+    # the sampled value (levels 1 to n - #(dropped levels)) to levels with random dropouts.
+    discreteVectorMapping = lapply(filterParams(par.set, type = c("discretevector", "logicalvector"))$pars,
+        function(param) rep(list(setNames(names(param$values), names(param$values))), param$len))
+    discreteVectorMapping = do.call(base::c, discreteVectorMapping)
```
Comment (jakob-r) [inline R/infillOptFocus.R]: see above

File: R/infillOptFocus.R [hunk 4]
```diff
@@ -8,11 +8,35 @@
...
+  # perform multiple starts
   for (restart.iter in seq_len(control$infill.opt.restarts)) {
     # copy parset so we can shrink it
     ps.local = par.set
+    
+    # handle discrete vectors:
+    # The problem is that for discrete vectors, we can't adjust the values dimension-wise.
+    # Therefore, for discrete vectors, we always drop the last level and instead have a
+    # mapping that maps, for each discrete vector param and for each dimension, from 
+    # the sampled value (levels 1 to n - #(dropped levels)) to levels with random dropouts.
```
Comment (jakob-r) [inline R/infillOptFocus.R]: I only maybe understand what you mean here :smiley:

File: R/infillOptFocus.R [hunk 5]
```diff
@@ -8,11 +8,35 @@
...
+    discreteVectorMapping = lapply(filterParams(par.set, type = c("discretevector", "logicalvector"))$pars,
+        function(param) rep(list(setNames(names(param$values), names(param$values))), param$len))
```
Comment (jakob-r) [inline R/infillOptFocus.R]: Maybe a comment would help "For each discrete/logic vector of length n create n replications of vector, each containing all possible values"?

File: R/infillOptFocus.R [hunk 6]
```diff
@@ -8,11 +8,35 @@
...
   for (restart.iter in seq_len(control$infill.opt.restarts)) {
     # copy parset so we can shrink it
     ps.local = par.set
+    
+    # handle discrete vectors:
+    # The problem is that for discrete vectors, we can't adjust the values dimension-wise.
+    # Therefore, for discrete vectors, we always drop the last level and instead have a
+    # mapping that maps, for each discrete vector param and for each dimension, from 
+    # the sampled value (levels 1 to n - #(dropped levels)) to levels with random dropouts.
+    discreteVectorMapping = lapply(filterParams(par.set, type = c("discretevector", "logicalvector"))$pars,
```
Comment (jakob-r) [inline R/infillOptFocus.R]: IMHO you can store `filterParams(par.set, type = c("discretevector", "logicalvector"))` as `ps.disic.log.vec` or something, as you use it twice here.

File: R/infillOptFocus.R (hunk 2 above)
Comment (mb706) [inline R/infillOptFocus.R]: `base::all.vars` returning a vector of all variables used in the expression? Unfortunately it is not perfect (ideally I would want a list of all _free_ variables) but it is the best R is giving me as it seems.

**Comment by mb706** [comment]
Incorporated your suggestions (hadn't thought of `unlist(..., recursive = FALSE)`, thanks) and tried to be more clear in the comment.

**Comment by mb706** [comment]
I don't think it is this PR's fault that the check fails.

**Comment by jakob-r** [comment]
I think we should pull this soon, or? @berndbischl

**Comment by berndbischl** [comment]
can somebody pls fix the naming issues? the styleguide is really followed. 

for this, please put this in a branch, we cannot do this in a fork. 
this seems to be an important bug fix.

i will proofread later then, please ping me.

and check 1000 million times if all reported problems from the issue have been converted to unit tests

**Comment by jakob-r** [comment]
Continue in #257

**Pull request closed without merging**